File: solana_toolbox_idl/src/toolbox_idl_type_flat_hydrate.rs
use std::collections::HashMap;
use std::sync::Arc;

use anyhow::anyhow;
use anyhow::Context;
use anyhow::Result;

use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlat;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatEnumVariant;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFieldNamed;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFieldUnnamed;
use crate::toolbox_idl_type_flat::ToolboxIdlTypeFlatFields;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFull;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullEnumVariant;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFieldNamed;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFieldUnnamed;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullFields;
use crate::toolbox_idl_type_full::ToolboxIdlTypeFullOrConstLiteral;
use crate::toolbox_idl_typedef::ToolboxIdlTypedef;
use crate::toolbox_idl_utils::idl_map_get_key_or_else;

impl ToolboxIdlTypeFlat {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFull> {
        match self.try_hydrate_or_const_literal(generics_by_symbol, typedefs)? {
            ToolboxIdlTypeFullOrConstLiteral::ToolboxIdlTypeFull(type_full) => {
                Ok(type_full)
            },
            ToolboxIdlTypeFullOrConstLiteral::ConstLiteral(literal) => Err(
                anyhow!("Const literal is not a standalone type: {}", literal),
            ),
        }
    }

    pub fn try_hydrate_or_const_literal(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullOrConstLiteral> {
        Ok(ToolboxIdlTypeFullOrConstLiteral::ToolboxIdlTypeFull(
            match self {
                ToolboxIdlTypeFlat::Defined { name, generics } => {
                    let typedef = idl_map_get_key_or_else(typedefs, name)
                        .context("Defined: Lookup")?;
                    if generics.len() < typedef.generics.len() {
                        return Err(anyhow!(
                        "Defined: Insufficient number of generic parameter for {}: expected: {}, found: {}",
                        typedef.name,
                        typedef.generics.len(),
                        generics.len()
                    ));
                    }
                    let generics_by_symbol = typedef
                        .generics
                        .iter()
                        .enumerate()
                        .map(|(index, generic_name)| {
                            generics
                                .get(index)
                                .unwrap()
                                .try_hydrate_or_const_literal(
                                    generics_by_symbol,
                                    typedefs,
                                )
                                .with_context(|| {
                                    format!(
                                        "Defined: {}, Generic[{}]",
                                        name, generic_name
                                    )
                                })
                                .map(|generic_type| {
                                    (generic_name.to_string(), generic_type)
                                })
                        })
                        .collect::<Result<
                            HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
                        >>()?;
                    let type_full = typedef
                        .type_flat
                        .try_hydrate(&generics_by_symbol, typedefs)
                        .with_context(|| {
                            format!("Defined: {}, Content", name)
                        })?;
                    match &typedef.serialization {
                        Some(serialization) if serialization == "bytemuck" => {
                            type_full
                                .bytemuck(&typedef.name, &typedef.repr)
                                .with_context(|| {
                                    format!("Defined: {}, Bytemuck", name)
                                })?
                                .2
                        },
                        _ => ToolboxIdlTypeFull::Typedef {
                            name: typedef.name.clone(),
                            repr: typedef.repr.clone(),
                            content: Box::new(type_full),
                        },
                    }
                },
                ToolboxIdlTypeFlat::Generic { symbol } => {
                    return Ok(idl_map_get_key_or_else(
                        generics_by_symbol,
                        symbol,
                    )
                    .with_context(|| format!("Generic Lookup: {}", symbol))?
                    .clone())
                },
                ToolboxIdlTypeFlat::Option { prefix, content } => {
                    ToolboxIdlTypeFull::Option {
                        prefix: prefix.clone(),
                        content: Box::new(
                            content
                                .try_hydrate(generics_by_symbol, typedefs)?,
                        ),
                    }
                },
                ToolboxIdlTypeFlat::Vec { prefix, items } => {
                    ToolboxIdlTypeFull::Vec {
                        prefix: prefix.clone(),
                        items: Box::new(
                            items.try_hydrate(generics_by_symbol, typedefs)?,
                        ),
                    }
                },
                ToolboxIdlTypeFlat::Array { items, length } => {
                    ToolboxIdlTypeFull::Array {
                        items: Box::new(
                            items.try_hydrate(generics_by_symbol, typedefs)?,
                        ),
                        length: match length.try_hydrate_or_const_literal(
                            generics_by_symbol,
                            typedefs,
                        )? {
                            ToolboxIdlTypeFullOrConstLiteral::ConstLiteral(
                                literal,
                            ) => {
                                usize::try_from(literal).with_context(|| {
                                    format!(
                                        "Array: length: invalid: {}",
                                        literal
                                    )
                                })?
                            },
                            _ => {
                                return Err(anyhow!(
                                    "Array: length must be a const literal"
                                ))
                            },
                        },
                    }
                },
                ToolboxIdlTypeFlat::String { prefix } => {
                    ToolboxIdlTypeFull::String {
                        prefix: prefix.clone(),
                    }
                },
                ToolboxIdlTypeFlat::Struct { fields } => {
                    ToolboxIdlTypeFull::Struct {
                        fields: fields
                            .try_hydrate(generics_by_symbol, typedefs)?,
                    }
                },
                ToolboxIdlTypeFlat::Enum { prefix, variants } => {
                    ToolboxIdlTypeFull::Enum {
                        prefix: prefix.clone(),
                        variants: variants
                            .iter()
                            .map(|variant| {
                                variant
                                    .try_hydrate(generics_by_symbol, typedefs)
                            })
                            .collect::<Result<Vec<_>, _>>()?,
                    }
                },
                ToolboxIdlTypeFlat::Padded {
                    before,
                    min_size,
                    after,
                    content,
                } => ToolboxIdlTypeFull::Padded {
                    before: *before,
                    min_size: *min_size,
                    after: *after,
                    content: Box::new(
                        content.try_hydrate(generics_by_symbol, typedefs)?,
                    ),
                },
                ToolboxIdlTypeFlat::Const { literal } => {
                    return Ok(ToolboxIdlTypeFullOrConstLiteral::ConstLiteral(
                        *literal,
                    ))
                },
                ToolboxIdlTypeFlat::Primitive { primitive } => {
                    ToolboxIdlTypeFull::Primitive {
                        primitive: primitive.clone(),
                    }
                },
            },
        ))
    }
}

impl ToolboxIdlTypeFlatEnumVariant {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullEnumVariant> {
        let variant_full_fields = self
            .fields
            .try_hydrate(generics_by_symbol, typedefs)
            .with_context(|| {
                format!("Variant: {}({})", self.name, self.code)
            })?;
        Ok(ToolboxIdlTypeFullEnumVariant {
            name: self.name.to_string(),
            code: self.code,
            fields: variant_full_fields,
        })
    }
}

impl ToolboxIdlTypeFlatFields {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFields> {
        Ok(match self {
            ToolboxIdlTypeFlatFields::Nothing => {
                ToolboxIdlTypeFullFields::Nothing
            },
            ToolboxIdlTypeFlatFields::Named(fields) => {
                let mut fields_full = vec![];
                for field in fields {
                    fields_full
                        .push(field.try_hydrate(generics_by_symbol, typedefs)?);
                }
                ToolboxIdlTypeFullFields::Named(fields_full)
            },
            ToolboxIdlTypeFlatFields::Unnamed(fields) => {
                let mut fields_type_full = vec![];
                for (index, field) in fields.iter().enumerate() {
                    fields_type_full.push(field.try_hydrate(
                        index,
                        generics_by_symbol,
                        typedefs,
                    )?);
                }
                ToolboxIdlTypeFullFields::Unnamed(fields_type_full)
            },
        })
    }
}

impl ToolboxIdlTypeFlatFieldNamed {
    pub fn try_hydrate(
        &self,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFieldNamed> {
        Ok(ToolboxIdlTypeFullFieldNamed {
            name: self.name.to_string(),
            content: self
                .content
                .try_hydrate(generics_by_symbol, typedefs)
                .with_context(|| format!("Named Field: {}", self.name))?,
        })
    }
}

impl ToolboxIdlTypeFlatFieldUnnamed {
    pub fn try_hydrate(
        &self,
        index: usize,
        generics_by_symbol: &HashMap<String, ToolboxIdlTypeFullOrConstLiteral>,
        typedefs: &HashMap<String, Arc<ToolboxIdlTypedef>>,
    ) -> Result<ToolboxIdlTypeFullFieldUnnamed> {
        Ok(ToolboxIdlTypeFullFieldUnnamed {
            position: index,
            content: self
                .content
                .try_hydrate(generics_by_symbol, typedefs)
                .with_context(|| format!("Unnamed Field: {}", index))?,
        })
    }
}

File: solana_toolbox_web3/src/ToolboxIdlTypeFlat.hydrate.ts
import { ToolboxIdlTypedef } from './ToolboxIdlTypedef';
import {
  ToolboxIdlTypeFlat,
  ToolboxIdlTypeFlatArray,
  ToolboxIdlTypeFlatConst,
  ToolboxIdlTypeFlatDefined,
  ToolboxIdlTypeFlatEnum,
  ToolboxIdlTypeFlatFieldNamed,
  ToolboxIdlTypeFlatFields,
  ToolboxIdlTypeFlatFieldUnnamed,
  ToolboxIdlTypeFlatGeneric,
  ToolboxIdlTypeFlatOption,
  ToolboxIdlTypeFlatPadded,
  ToolboxIdlTypeFlatString,
  ToolboxIdlTypeFlatStruct,
  ToolboxIdlTypeFlatVec,
} from './ToolboxIdlTypeFlat';
import {
  ToolboxIdlTypeFull,
  ToolboxIdlTypeFullFields,
} from './ToolboxIdlTypeFull';
import { bytemuck } from './ToolboxIdlTypeFull.bytemuck';
import { ToolboxIdlTypePrimitive } from './ToolboxIdlTypePrimitive';

export function hydrate(
  typeFlat: ToolboxIdlTypeFlat,
  genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
  typedefs: Map<string, ToolboxIdlTypedef>,
): ToolboxIdlTypeFull {
  const typeFullOrConstLiteral = hydrateOrConstLiteral(
    typeFlat,
    genericsBySymbol,
    typedefs,
  );
  if (typeof typeFullOrConstLiteral === 'number') {
    throw new Error('Const is not supported as a standalone type');
  }
  return typeFullOrConstLiteral;
}

export function hydrateOrConstLiteral(
  typeFlat: ToolboxIdlTypeFlat,
  genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
  typedefs: Map<string, ToolboxIdlTypedef>,
): ToolboxIdlTypeFull | number {
  return typeFlat.traverse(
    hydrateOrConstLiteralVisitor,
    genericsBySymbol,
    typedefs,
  );
}

let hydrateOrConstLiteralVisitor = {
  defined: (
    self: ToolboxIdlTypeFlatDefined,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    const typedef = typedefs.get(self.name);
    if (typedef === undefined) {
      throw new Error(`Could not resolve type named: ${self.name}`);
    }
    if (self.generics.length < typedef.generics.length) {
      throw new Error('Insufficient set of generics');
    }
    const genericsFull = self.generics.map(
      (genericFlat: ToolboxIdlTypeFlat) => {
        return hydrateOrConstLiteral(genericFlat, genericsBySymbol, typedefs);
      },
    ); // TODO - this could be safer and cleaner by using strong types
    const innerGenericsBySymbol = new Map<
      string,
      ToolboxIdlTypeFull | number
    >();
    for (let i = 0; i < typedef.generics.length; i++) {
      innerGenericsBySymbol.set(typedef.generics[i]!, genericsFull[i]!);
    }
    const typeFull = hydrate(typedef.typeFlat, innerGenericsBySymbol, typedefs);
    const typeTypedef = {
      name: typedef.name,
      repr: typedef.repr,
      content: typeFull,
    };
    if (typedef.serialization === 'bytemuck') {
      return bytemuck(typeTypedef).value;
    }
    return ToolboxIdlTypeFull.typedef(typeTypedef);
  },
  generic: (
    self: ToolboxIdlTypeFlatGeneric,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    let typeFull = genericsBySymbol.get(self.symbol);
    if (typeFull === undefined) {
      throw new Error(`Could not resolve generic named: ${self.symbol}`);
    }
    return typeFull;
  },
  option: (
    self: ToolboxIdlTypeFlatOption,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.option({
      prefix: self.prefix,
      content: hydrate(self.content, genericsBySymbol, typedefs),
    });
  },
  vec: (
    self: ToolboxIdlTypeFlatVec,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.vec({
      prefix: self.prefix,
      items: hydrate(self.items, genericsBySymbol, typedefs),
    });
  },
  array: (
    self: ToolboxIdlTypeFlatArray,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    const length = hydrateOrConstLiteral(
      self.length,
      genericsBySymbol,
      typedefs,
    );
    if (typeof length !== 'number') {
      throw new Error('Array length must resolve to a const literal number');
    }
    return ToolboxIdlTypeFull.array({
      length,
      items: hydrate(self.items, genericsBySymbol, typedefs),
    });
  },
  string: (
    self: ToolboxIdlTypeFlatString,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.string({
      prefix: self.prefix,
    });
  },
  struct: (
    self: ToolboxIdlTypeFlatStruct,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.struct({
      fields: hydrateFields(self.fields, genericsBySymbol, typedefs),
    });
  },
  enum: (
    self: ToolboxIdlTypeFlatEnum,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.enum({
      prefix: self.prefix,
      variants: self.variants.map((variant) => {
        return {
          name: variant.name,
          code: variant.code,
          fields: hydrateFields(variant.fields, genericsBySymbol, typedefs),
        };
      }),
    });
  },
  padded: (
    self: ToolboxIdlTypeFlatPadded,
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.padded({
      before: self.before,
      minSize: self.minSize,
      after: self.after,
      content: hydrate(self.content, genericsBySymbol, typedefs),
    });
  },
  const: (
    self: ToolboxIdlTypeFlatConst,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return self.literal;
  },
  primitive: (
    self: ToolboxIdlTypePrimitive,
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFull | number => {
    return ToolboxIdlTypeFull.primitive(self);
  },
};

export function hydrateFields(
  typeFlatFields: ToolboxIdlTypeFlatFields,
  genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
  typedefs: Map<string, ToolboxIdlTypedef>,
): ToolboxIdlTypeFullFields {
  return typeFlatFields.traverse(
    hydrateFieldsVisitor,
    genericsBySymbol,
    typedefs,
  );
}

let hydrateFieldsVisitor = {
  nothing: (
    _self: {},
    _genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    _typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFullFields => {
    return ToolboxIdlTypeFullFields.nothing();
  },
  named: (
    self: ToolboxIdlTypeFlatFieldNamed[],
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFullFields => {
    return ToolboxIdlTypeFullFields.named(
      self.map((field) => {
        return {
          name: field.name,
          content: hydrate(field.content, genericsBySymbol, typedefs),
        };
      }),
    );
  },
  unnamed: (
    self: ToolboxIdlTypeFlatFieldUnnamed[],
    genericsBySymbol: Map<string, ToolboxIdlTypeFull | number>,
    typedefs: Map<string, ToolboxIdlTypedef>,
  ): ToolboxIdlTypeFullFields => {
    return ToolboxIdlTypeFullFields.unnamed(
      self.map((field, index) => {
        return {
          position: index,
          content: hydrate(field.content, genericsBySymbol, typedefs),
        };
      }),
    );
  },
};
